### PR TITLE
[Decoder/DirectVideo] add the video format RGBA and so on @open sesame 04/06 16:36

### PR DESCRIPTION
**Pull request by helloahn**
To handle various cases, add an available format for the direct video decoder and add an option to handle it.

The example of the new option
```bash
gst-launch-1.0 \
...
        tensor_converter ! tensor_decoder mode=direct_video option1=RGBA ! \
...
```

Signed-off-by: Hyoung Joo Ahn <hello.ahn@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 1]
```diff
@@ -40,11 +41,45 @@ void fini_dv (void) __attribute__ ((destructor));
...
+};
+
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static int
 dv_init (void **pdata)
 {
-  *pdata = NULL;                /* We have no internal data */
+  direct_video_ops *ddata;
+
+  ddata = *pdata = g_new0 (direct_video_ops, 1);
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:77]: Better to use `g_try_new0`

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 2]
```diff
@@ -60,14 +95,31 @@ dv_exit (void **pdata)
...
 {
-  /* We do not accept anything. */
+  direct_video_ops *ddata = *pdata;
+
+  /* When the dimension[0] is 4, the video format will be decided by option1. */
+  switch (opNum) {
+    case 0:
+      ddata->format = find_key_strv (dv_formats, param);
+
+      if (NULL == param || *param == '\0') {
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:105]: As `param` is const type, I think this statement should be moved to before `find_key_strv`

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 1 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: looks better

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 2 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:117]: that's right!

**Review by dongju-chae — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 3]
```diff
@@ -40,11 +41,59 @@ void fini_dv (void) __attribute__ ((destructor));
...
+{
+  DIRECT_VIDEO_FORMAT_RGBx = 0,
+  DIRECT_VIDEO_FORMAT_BGRx = 1,
+  DIRECT_VIDEO_FORMAT_xRGB = 2,
+  DIRECT_VIDEO_FORMAT_xBGR = 3,
+  DIRECT_VIDEO_FORMAT_RGBA = 4,
+  DIRECT_VIDEO_FORMAT_BGRA = 5,
+  DIRECT_VIDEO_FORMAT_ARGB = 6,
+  DIRECT_VIDEO_FORMAT_ABGR = 7,
+  DIRECT_VIDEO_FORMAT_I420 = 8
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:57]: I420 is improper

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 4]
```diff
@@ -88,7 +153,24 @@ dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
       format = GST_VIDEO_FORMAT_RGB;
       break;
     case 4:
-      format = GST_VIDEO_FORMAT_BGRx;
+      if (ddata->format == DIRECT_VIDEO_FORMAT_RGBx)
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:156]: better changing switch (ddata->format)..case

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 5]
```diff
@@ -40,11 +41,59 @@ void fini_dv (void) __attribute__ ((destructor));
     GST_VIDEO_CAPS_MAKE ("{ RGB, BGRx, GRAY8 }") \
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:41]: when adding additional video format, you should update this macro.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 6]
```diff
@@ -40,11 +41,59 @@ void fini_dv (void) __attribute__ ((destructor));
...
+};
+
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static int
 dv_init (void **pdata)
 {
-  *pdata = NULL;                /* We have no internal data */
+  direct_video_ops *ddata;
+
+  ddata = *pdata = g_try_new0 (direct_video_ops, 1);
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:91]: Where is this released?

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 3 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: it's moved to channel 1 and other formats have been updated

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 4 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: it's applied

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 5 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: it's added

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 6 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:104]: it's added at the exit function.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 7]
```diff
@@ -80,19 +159,66 @@ dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
...
+      case DIRECT_VIDEO_FORMAT_I420:
+        format = GST_VIDEO_FORMAT_I420;
+        break;
+      case DIRECT_VIDEO_FORMAT_GRAY8:
+        format = GST_VIDEO_FORMAT_GRAY8;
+        break;
+      default:
+        GST_ERROR ("Invalid Video Format. With 1 channel, only these formats are available: I420, GRAY8");
+        return NULL;
+        break;
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:174]: minor: Unnecessary break

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 8]
```diff
@@ -80,19 +159,66 @@ dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
...
+      case DIRECT_VIDEO_FORMAT_RGB:
+        format = GST_VIDEO_FORMAT_RGB;
+        break;
+      case DIRECT_VIDEO_FORMAT_BGR:
+        format = GST_VIDEO_FORMAT_BGR;
+        break;
+      default:
+        GST_ERROR ("Invalid Video Format. With 3 channels, only these formats are available: RGB, BGR");
+        return NULL;
+        break;
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:187]: ditto

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 9]
```diff
@@ -30,47 +30,126 @@
...
+  /* From option1 */
+  direct_video_formats format;
+} direct_video_ops;
+
+/**
+ * @brief List of the formats of direct video
+ */
+static const char *dv_formats[] = {
+  [DIRECT_VIDEO_FORMAT_UNKNOWN] = "UNKNOWN",
+  [DIRECT_VIDEO_FORMAT_I420] = "I420",
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:84]: Q. Does tensor support this type (I420)?

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 10]
```diff
@@ -80,19 +159,66 @@ dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
...
+  } else if (channel == 3) {
+    switch (ddata->format) {
+      case DIRECT_VIDEO_FORMAT_RGB:
+        format = GST_VIDEO_FORMAT_RGB;
+        break;
+      case DIRECT_VIDEO_FORMAT_BGR:
+        format = GST_VIDEO_FORMAT_BGR;
+        break;
+      default:
+        GST_ERROR ("Invalid Video Format. With 3 channels, only these formats are available: RGB, BGR");
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:185]: if channel is 3, default GST_VIDEO_FORMAT_RGB if no option param is given.

Comment (myungjoo) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:185]: Besides, it should be able to distinguish between "unspecified" and "others (error)".

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 7 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:182]: now it is :)

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 9 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: it is impossible with rank 4 currently :) it's removed.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 10 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: I set the default to UNKNOWN, and if it is, the format will be set to the default according to the number of channels.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 11]
```diff
@@ -30,47 +30,131 @@
...
+};
+
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static int
 dv_init (void **pdata)
 {
-  *pdata = NULL;                /* We have no internal data */
+  direct_video_ops *ddata;
+  ddata = *pdata = g_try_new0 (direct_video_ops, 1);
+  ddata->format = DIRECT_VIDEO_FORMAT_UNKNOWN;
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:103]: This line should be located after NULL check.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 12]
```diff
@@ -30,47 +30,131 @@
...
+  }
+  
   return TRUE;
 }
 
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static GstCaps *
 dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
 {
+  direct_video_ops *ddata = *pdata;
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:154]: Let's check if ddata is null or not.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 11 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: nice catch :+1:

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 12 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:158]: it's added at line:134

**Review by gichan-jang — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 13]
```diff
@@ -80,19 +168,67 @@ dv_getOutCaps (void **pdata, const GstTensorsConfig * config)
   caps = gst_caps_from_string (DECODER_DV_VIDEO_CAPS_STR);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:168]: let's move this to line 236

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 14]
```diff
@@ -30,74 +30,216 @@
...
-    default:
-      format = GST_VIDEO_FORMAT_UNKNOWN;
-      break;
+  channel = config->info.info[0].dimension[0];
+  if (channel == 1) {
+    switch (ddata->format) {
+      case DIRECT_VIDEO_FORMAT_GRAY8:
+        format = GST_VIDEO_FORMAT_GRAY8;
+        break;
+      case DIRECT_VIDEO_FORMAT_UNKNOWN:
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:181]: What I'd prefer is:
```
  case DIRECT_VIDEO_FORMAT_UNKNOWN:
    format = GST_VIDEO_FORMAT_GRAY8;
    break;
  default:
    HANDLE ERROR
```

>  distinguish between "unspecified (use default)" and "others (error)".

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c [hunk 15]
```diff
@@ -30,74 +30,216 @@
...
+
+  ddata = *pdata;
+
+  /* When the dimension[0] is 4, the video format will be decided by option1. */
+  switch (opNum) {
+    case 0:
+    {
+      int f = find_key_strv (dv_formats, param);
+
+      ddata->format = (f < 0) ? f : DIRECT_VIDEO_FORMAT_UNKNOWN;
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:146]: oh...my fault. this is invalid condition.

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 14 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c:181]: it has been divided

File: ext/nnstreamer/tensor_decoder/tensordec-directvideo.c (hunk 15 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_decoder/tensordec-directvideo.c]: fixed!

**Review by myungjoo — APPROVED** [review]
LGTM

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**